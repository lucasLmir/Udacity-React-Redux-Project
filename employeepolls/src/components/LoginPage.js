import { connect } from "react-redux";
import { useState } from "react";
import { setAuthedUser } from "../actions/authedUser";
import { useNavigate } from "react-router-dom";

const LoginPage = (props) => {
  const [uname, setUname] = useState("");
  const [pass, setPass] = useState("");
  const [message, setMessage] = useState("");
  const navigate = useNavigate();

  const handleSubmit = (e) => {
    e.preventDefault();
    const match = props.users[uname].password === pass;

    if (match) {
      props.dispatch(setAuthedUser(uname));
      navigate("/");
    } else {
      setMessage("Username or Password incorrect!");
    }
  };

  const handleLogout = (e) => {
    props.dispatch(setAuthedUser(null));
  };

  return (
    <div>
      <button
        data-testid="logout-btn"
        hidden={props.authedUser === null}
        onClick={handleLogout}
      >
        LogOut
      </button>
      <form onSubmit={handleSubmit} data-testid="login-form">
        <input
          data-testid="uname-input"
          type="text"
          value={uname}
          onChange={(e) => setUname(e.target.value)}
          name="Uname"
          id="Uname"
          placeholder="Username"
          hidden={props.authedUser !== null}
        />
        <br />
        <input
          data-testid="pass-input"
          type="Password"
          value={pass}
          onChange={(e) => setPass(e.target.value)}
          name="Pass"
          id="Pass"
          placeholder="Password"
          hidden={props.authedUser !== null}
        />
        <br />
        <button data-testid="submit-btn" hidden={props.authedUser !== null}>LogIn</button>
      </form>
      <p data-testid="message">{message}</p>
    </div>
  );
};

const mapStateToProps = ({ authedUser, users }) => {
  return {
    users,
    authedUser,
  };
};
export default connect(mapStateToProps)(LoginPage);
