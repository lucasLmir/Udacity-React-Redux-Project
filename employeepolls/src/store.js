import reducer from "./reducers";
import { createStore } from "redux";
import middleware from "./middleware";

export const store = createStore(reducer, middleware);