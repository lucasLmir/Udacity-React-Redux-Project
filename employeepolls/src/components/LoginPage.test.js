import { render, fireEvent } from "@testing-library/react";
import LoginPage from "./LoginPage";
import "@testing-library/jest-dom";
import { store } from "../store";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import { setAuthedUser } from "../actions/authedUser";
import { receiveUsers } from "../actions/users";

describe("LoginPage", () => {
  it("is expected to render the component", () => {
    const component = render(
      <Provider store={store}>
        <Router>
          <LoginPage />
        </Router>
      </Provider>
    );
    expect(component).toMatchSnapshot();
  });

  it("Is expected to show an error message if a wrong password is typed in the password input field.", async () => {
    const users = {
      mtsamis: {
        id: 'mtsamis',
        password:'xyz123',
        name: 'Mike Tsamis',
        avatarURL: "https://www.shutterstock.com/image-vector/young-smiling-man-adam-avatar-600w-2107967969.jpg",
      }
    }
    await store.dispatch(receiveUsers(users));
    const component = render(
      <Provider store={store}>
        <Router>
          <LoginPage />
        </Router>
      </Provider>
    );

    const unameInput = component.getByTestId("uname-input");
    fireEvent.change(unameInput, { target: { value: "mtsamis" } });
    const passInput = component.getByTestId("pass-input");
    fireEvent.change(passInput, { target: { value: "notpassword" } });
    const submitButton = component.getByTestId("submit-btn");
    fireEvent.click(submitButton);
    expect(component.getByTestId("message").innerHTML).toBe("Username or Password incorrect!");
    expect(component.queryByTestId("submit-btn")).toBeInTheDocument();
  });

  it("It is expected to show log out button if a user is logged in.", async () => {
    const user = "mtsamis";
    await store.dispatch(setAuthedUser(user));
    const component = render(
      <Provider store={store}>
        <Router>
          <LoginPage />
        </Router>
      </Provider>
    );
    expect(component.queryByTestId("logout-btn")).toBeInTheDocument();
  });
});
