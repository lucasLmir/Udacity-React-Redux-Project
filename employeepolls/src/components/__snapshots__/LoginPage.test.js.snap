// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage is expected to render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <button
          data-testid="logout-btn"
          hidden=""
        >
          LogOut
        </button>
        <form
          data-testid="login-form"
        >
          <input
            data-testid="uname-input"
            id="Uname"
            name="Uname"
            placeholder="Username"
            type="text"
            value=""
          />
          <br />
          <input
            data-testid="pass-input"
            id="Pass"
            name="Pass"
            placeholder="Password"
            type="Password"
            value=""
          />
          <br />
          <button
            data-testid="submit-btn"
          >
            LogIn
          </button>
        </form>
        <p
          data-testid="message"
        />
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <button
        data-testid="logout-btn"
        hidden=""
      >
        LogOut
      </button>
      <form
        data-testid="login-form"
      >
        <input
          data-testid="uname-input"
          id="Uname"
          name="Uname"
          placeholder="Username"
          type="text"
          value=""
        />
        <br />
        <input
          data-testid="pass-input"
          id="Pass"
          name="Pass"
          placeholder="Password"
          type="Password"
          value=""
        />
        <br />
        <button
          data-testid="submit-btn"
        >
          LogIn
        </button>
      </form>
      <p
        data-testid="message"
      />
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
